File: tests/test_cost_train.py
"""
equivalent test of run-cost-train.sh in
https://github.com/taku910/mecab/blob/master/mecab/tests/run-cost-train.sh
"""

import os

import pytest

from mecab.cli import (
    run_mecab_cost_train,
    run_mecab_dict_gen,
    run_mecab_dict_index,
    run_mecab_main,
    run_mecab_system_eval,
    run_mecab_test_gen,
)

training_result = """
              precision          recall         F
LEVEL 0:    12.4183(57/459) 11.8998(57/479) 12.1535
LEVEL 1:    11.7647(54/459) 11.2735(54/479) 11.5139
LEVEL 2:    11.3290(52/459) 10.8559(52/479) 11.0874
LEVEL 4:    11.3290(52/459) 10.8559(52/479) 11.0874"""

# test multithreaded training
@pytest.mark.parametrize("threadnum", [pytest.param("1"), pytest.param("4")])
def test_cost_train(threadnum, tmpdir, capfd):
    """check parameter estimation is possible"""
    COST_TRAIN_SEED_PATH = "tests/test-data/cost-train/seed"
    CORPUS_PATH = "tests/test-data/cost-train/training-data.txt"
    COST_TRAIN_TEST_PATH = "tests/test-data/cost-train/test-data.txt"
    MODEL_PATH = str(tmpdir.join("model.bin"))
    DICTIONARY_PATH = str(tmpdir.mkdir("dic"))
    TEST_PROCESSED_PATH = str(tmpdir.join("test.txt"))
    EVAL_RESULT_PATH = str(tmpdir.join("result.txt"))

    assert os.path.exists(CORPUS_PATH)
    assert os.path.exists(COST_TRAIN_SEED_PATH)
    assert os.path.exists(COST_TRAIN_TEST_PATH)

    # move dicrc
    for filename in os.listdir(COST_TRAIN_SEED_PATH):
        _copy_file(os.path.join(COST_TRAIN_SEED_PATH, filename), tmpdir.join(filename))

    run_mecab_dict_index(["index", "-d", COST_TRAIN_SEED_PATH, "-o", str(tmpdir)])
    run_mecab_cost_train(["train", "-c", "1.0", "-p", threadnum, "-d", str(tmpdir), "-f", "1", CORPUS_PATH, MODEL_PATH])
    run_mecab_dict_gen(["dgen", "-d", str(tmpdir), "-m", MODEL_PATH, "-o", DICTIONARY_PATH])
    run_mecab_dict_index(["index", "-d", DICTIONARY_PATH, "-o", DICTIONARY_PATH])
    run_mecab_test_gen(["tgen", "-o", TEST_PROCESSED_PATH, COST_TRAIN_TEST_PATH])
    run_mecab_main(["mecab", "-r", "/dev/null", "-d", DICTIONARY_PATH, "-o", EVAL_RESULT_PATH, TEST_PROCESSED_PATH])
    run_mecab_system_eval(["eval", "-l", "0 1 2 4", EVAL_RESULT_PATH, COST_TRAIN_TEST_PATH])

    captured = capfd.readouterr().out
    assert training_result in captured


def _copy_file(from_, to):
    with open(from_) as f:
        to.write(f.read())

File: tests-integration/test_cost_train.cc
#include <fstream>

#include "fixture/tmpdir.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "mecab/cli.h"

#define MAKE_ARGS(arg_name, ...)                                   \
  std::vector<std::string> __arguments{__VA_ARGS__};               \
  std::vector<char*> arg_name;                                     \
  for (size_t i = 0; i < __arguments.size(); ++i) {                \
    arg_name.push_back(const_cast<char*>(__arguments[i].c_str())); \
  }

namespace {
bool is_exists(std::string path) {
  struct stat info;
  return stat(path.c_str(), &info) == 0;
}

void copy_file(std::string from, std::string to) {
  std::ifstream src(from, std::ios::binary);
  std::ofstream dst(to, std::ios::binary);

  dst << src.rdbuf();
}
}  // namespace

/**
 * equivalent test of run-cost-train.sh in
 * https://github.com/taku910/mecab/blob/master/mecab/tests/run-cost-train.sh
 */
TEST(mecab_integration_test, cost_train) {
  fixture::TmpDir tmpdir;

  const std::string trainingResult =
      "              precision          recall         F\n"
      "LEVEL 0:    12.4183(57/459) 11.8998(57/479) 12.1535\n"
      "LEVEL 1:    11.7647(54/459) 11.2735(54/479) 11.5139\n"
      "LEVEL 2:    11.3290(52/459) 10.8559(52/479) 11.0874\n"
      "LEVEL 4:    11.3290(52/459) 10.8559(52/479) 11.0874";
  const std::string corpusPath = "../test-data/cost-train/training-data.txt";
  const std::string costTrainSeedPath = "../test-data/cost-train/seed";
  const std::string costTrainTestPath = "../test-data/cost-train/test-data.txt";

  ASSERT_TRUE(is_exists(corpusPath));
  ASSERT_TRUE(is_exists(costTrainSeedPath));
  ASSERT_TRUE(is_exists(costTrainTestPath));

  const std::string modelPath = tmpdir.getPath() + "/model.bin";
  const std::string dicPath = tmpdir.createPath("dic");
  const std::string testProcessedPath = tmpdir.getPath() + "/test.txt";
  const std::string evalResultPath = tmpdir.getPath() + "/result.txt";

  for (auto file : {"char.def", "dicrc", "dictionary.csv", "feature.def", "rewrite.def", "unk.def"}) {
    copy_file(costTrainSeedPath + "/" + file, tmpdir.getPath() + "/" + file);
  }

  testing::internal::CaptureStdout();
  {
    MAKE_ARGS(mecab_dict_index_args, "mecab-dict-index", "-d", costTrainSeedPath, "-o", tmpdir.getPath());
    mecab_dict_index(mecab_dict_index_args.size(), mecab_dict_index_args.data());
  }
  {
    MAKE_ARGS(mecab_cost_train_args, "mecab-cost-train", "-c", "1.0", "-p", "1", "-d", tmpdir.getPath(), "-f", "1",
              corpusPath, modelPath);
    mecab_cost_train(mecab_cost_train_args.size(), mecab_cost_train_args.data());
  }
  {
    MAKE_ARGS(mecab_dict_gen_args, "mecab-dict-gen", "-d", tmpdir.getPath(), "-m", modelPath, "-o", dicPath);
    mecab_dict_gen(mecab_dict_gen_args.size(), mecab_dict_gen_args.data());
  }
  {
    MAKE_ARGS(mecab_dict_index_args, "mecab-dict-index", "-d", dicPath, "-o", dicPath);
    mecab_dict_index(mecab_dict_index_args.size(), mecab_dict_index_args.data());
  }
  {
    MAKE_ARGS(mecab_test_gen_args, "mecab-test-gen", "-o", testProcessedPath, costTrainTestPath);
    mecab_test_gen(mecab_test_gen_args.size(), mecab_test_gen_args.data());
  }
  {
    MAKE_ARGS(run_mecab_main_args, "mecab", "-r", "/dev/null", "-d", dicPath, "-o", evalResultPath, testProcessedPath);
    mecab_main(run_mecab_main_args.size(), run_mecab_main_args.data());
  }
  {
    MAKE_ARGS(mecab_system_eval_args, "mecab-system-eval", "-l", "0 1 2 4", evalResultPath, costTrainTestPath);
    mecab_system_eval(mecab_system_eval_args.size(), mecab_system_eval_args.data());
  }

  std::string captured = testing::internal::GetCapturedStdout();
  ASSERT_THAT(captured, ::testing::HasSubstr(trainingResult));
}
